doc/Laika_Architecture.md

```diff
@@ -0,0 +1,45 @@
+
+`PageDataInitializer` extracts the `selectedBuckets` variable from the application variables and puts it into a separate property. Use this variable as an input for the [a/b testing features (CSS, Vue, Store) in the client-side code](HOWTO_Create_an_a_b_test.md).
+
+`PageDataInitializer` also reads two other data attributes from the div:
+
+* `messages` - contains all i18n messages from the `fundraising-frontend-content` repository
+* `assetsPath` - path to assets (JS, CSS, images, fonts) of the laika skin
+
+## Component hierarchy, reuse and using the store
+When thinking about modularization of components, apply the following rules of thumb:
+
+* If a root component becomes too large, you should break it up into smaller components. Put them in a subdirectory in `laika/src/components/pages`, named after the root component (in snake_case). 
+* Keep the communication with the store (through the `mapState` and `mapGetters` helper functions and the `$store.dispatch` method) in the root component.
+* If a page (root component) has "subpages" that don't correspond to a PHP route, put them in a subdirectory following the naming schema `laika/src/pages/root_component_name/subpages`. Move the communication with the store to the subpages, the root component should handle the display logic for the subpages.
+* If you reuse components across pages, put them in the `laika/src/components/shared` directory. Shared components are not allowed to communicate with the store! If you catch yourself adding `if` statements for different use cases to shared components, move them back to the root component subdirectory and live with the duplication. Adding features that are not used by all parent components is ok, though.
+* Only add new state to the store if that state needs to be shared across child components.
+* Always dispatch actions to the store, never use mutations from the components! This ensures a consistent asynchronous interface. If you need to wrap mutations in actions for that, do it.
+
+
+## Possible for improvements for the Laika architecture
+* Investigate build speed - [T241851](https://phabricator.wikimedia.org/T241851)
+* Build CSS only once, not for every entry point.
+* Increase Type safety:
+  * remove `any` and casts.
+  * Better type validation of server-side data in  `PageDataInitializer`, maybe using https://github.com/joanllenas/ts.data.json or https://github.com/gcanti/io-ts
+  * Use Vue 3.0 composition api
+  * Rethink state management: Either make Vuex more type safe by using [vuex-class](https://github.com/ktsn/vuex-class) or [vuex-module-decorators](https://github.com/championswimmer/vuex-module-decorators) or gradually replace Vuex with a different reactive state managemnt solution that is a better fit for the composition API. Maybe [use the composition API for state management](https://medium.com/everything-full-stack/vue-composition-api-as-a-state-management-cbc509f5a717).
```

Comment:
The type safety of Vuex has been a concern in the Wikidata team as well. The Data Bridge feature team decided to replace vuex with [vuex-smart-modules](url) Here's the [related ADR](https://github.com/wikimedia/mediawiki-extensions-Wikibase/blob/master/docs/adr/0007-use-vuex-smart-modules.md) and the [Expected Consequences](https://github.com/wikimedia/mediawiki-extensions-Wikibase/blob/master/docs/adr/0007-use-vuex-smart-modules.md#expected-consequences) part is especially important.
The usage of Vue 3.0 is a great idea, as this version will give full TS support, aka no more awkward decorators.

This looks interesting https://github.com/posva/pinia but unfortunately is only experimental.
The latest unstable Vuex release hints at TypeScript support - https://github.com/vuejs/vuex/releases/tag/v4.0.0-alpha.1
As far as I understand, Vuex 4 aims to support Vue 3, therefore TS out of the box.